### ISSUE TITLE
Gem install skylight fails

### ISSUE THREAD

**Issue by hpetru** [skylightio/skylight-ruby#124]
Hello, I have this issue when I'm trying to install skylight

That's happens after I start using a version of forked rails and installing the rails gem from github

```
bundle install
```

```
Fetching gem metadata from https://rubygems.org/...........
Fetching gem metadata from https://rubygems.org/..
Resolving dependencies......
...
Using activesupport 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
...
Using loofah 2.2.2
Using rails-html-sanitizer 1.0.4
Using actionview 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
Using rack 1.6.10
Using rack-test 0.6.3
Using actionpack 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
Using globalid 0.4.1
Using activejob 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
Using mini_mime 1.0.0
Using mail 2.7.0
Using actionmailer 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
Using public_suffix 3.0.2
Using addressable 2.5.2
Using active_link_to 1.0.5
Using activemodel 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
Using active_model_serializers 0.8.3
Using arel 6.0.4
Using activerecord 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
Using multi_json 1.13.1
Using thor 0.20.0
Using railties 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
...
Using analytics-ruby 2.2.2
Using annotate 2.7.1 from https://github.com/ctran/annotate_models.git (at develop@625e0d0)
Using ansi 1.5.0
...
Using rails 4.2.10 from git@github.com:XXXXX/rails.git (at fixed/v4.2.10@f01074e)
Using page_title_helper 2.1.0
....
Fetching skylight 2.0.2
Installing skylight 2.0.2 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /Users/petru/.rvm/gems/ruby-2.3.0@XXXX/gems/skylight-2.0.2/ext
/Users/petru/.rvm/rubies/ruby-2.3.0/bin/ruby -r ./siteconf20180709-11961-90gjq5.rb extconf.rb
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/Users/petru/.rvm/rubies/ruby-2.3.0/bin/$(RUBY_BASE_NAME)
/Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/dependency.rb:308:in `to_specs': Could not find 'activesupport' (>= 4.2.0) among 402 total gem(s) (Gem::MissingSpecError)
Checked in 'GEM_PATH=/Users/petru/.rvm/gems/ruby-2.3.0@XXXX:/Users/petru/.rvm/gems/ruby-2.3.0@global', execute `gem env` for more information
	from /Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/specification.rb:1441:in `block in activate_dependencies'
	from /Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/specification.rb:1430:in `each'
	from /Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/specification.rb:1430:in `activate_dependencies'
	from /Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/specification.rb:1412:in `activate'
	from /Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_gem.rb:68:in `block in gem'
	from /Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_gem.rb:67:in `synchronize'
	from /Users/petru/.rvm/rubies/ruby-2.3.0/lib/ruby/site_ruby/2.3.0/rubygems/core_ext/kernel_gem.rb:67:in `gem'
	from extconf.rb:17:in `<main>'

extconf failed, exit code 1

Gem files will remain installed in /Users/petru/.rvm/gems/ruby-2.3.0@XXXX/gems/skylight-2.0.2 for inspection.
Results logged to /Users/petru/.rvm/gems/ruby-2.3.0@XXXX/extensions/x86_64-darwin-15/2.3.0/skylight-2.0.2/gem_make.out

An error occurred while installing skylight (2.0.2), and Bundler cannot continue.
Make sure that `gem install skylight -v '2.0.2'` succeeds before bundling.

In Gemfile:
  skylight

```

**Comment by wagenet** [comment]
It looks like the issue may be around Skylight not recognizing ActiveSupport. This could be due to your forked Rails.

**Comment by hpetru** [comment]
@wagenet `This could be due to your forked Rails.` yes, that's the reason, do you have any ideas how I can solve it ?

**Comment by hpetru** [comment]
@wagenet Yes, right, here is an example Gemfile which is not working too, getting same error
```
source "https://rubygems.org"

git 'https://github.com/hpetru/rails.git', tag: 'fix/statement-cache-issue' do
  gem 'rails'
  gem 'activesupport'
end

gem 'skylight'
```

**Comment by wagenet** [comment]
@hpetru Can you share the contents of ` /Users/petru/.rvm/gems/ruby-2.3.0@XXXX/gems/skylight-2.0.2/ext/mkmf.log`?

**Comment by wagenet** [comment]
@hpetru ok, just the same error. AFAICT, this is a Bundler issue and I'm not really certain what to do about it. I don't think we're doing anything unusual that would be causing this trouble.

**Comment by treycaliva** [comment]
@wagenet Having a similar issue. Skylight gem installs fine locally, but fails on deploy to Heroku.

**Comment by zvkemp** [comment]
@treycaliva Can you please provide your Heroku build output showing the failure?

**Comment by treycaliva** [comment]
`remote:        Installing skylight 2.0.2 with native extensions
remote:        Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
remote:
remote:        current directory:
remote:        /tmp/build_72bc4f42ff1f9778d777fc04914fb767/vendor/bundle/ruby/2.5.0/gems/skylight-2.0.2/ext
remote:        /tmp/build_72bc4f42ff1f9778d777fc04914fb767/vendor/ruby-2.5.1/bin/ruby -r
remote:        ./siteconf20180725-218-kvviww.rb extconf.rb
remote:        *** extconf.rb failed ***
remote:        Could not create Makefile due to some reason, probably lack of necessary
remote:        libraries and/or headers.  Check the mkmf.log file for more details.  You may
remote:        need configuration options.
remote:
remote:        Provided configuration options:
remote:        	--with-opt-dir
remote:        	--without-opt-dir
remote:        	--with-opt-include
remote:        	--without-opt-include=${opt-dir}/include
remote:        	--with-opt-lib
remote:        	--without-opt-lib=${opt-dir}/lib
remote:        	--with-make-prog
remote:        	--without-make-prog
remote:        	--srcdir=.
remote:        	--curdir
remote:        --ruby=/tmp/build_72bc4f42ff1f9778d777fc04914fb767/vendor/ruby-2.5.1/bin/$(RUBY_BASE_NAME)
remote:        /tmp/build_72bc4f42ff1f9778d777fc04914fb767/vendor/ruby-2.5.1/lib/ruby/2.5.0/rubygems/dependency.rb:310:in
remote:        `to_specs': Could not find 'activesupport' (>= 4.2.0) among 100 total gem(s)
remote:        (Gem::MissingSpecError)`

**Comment by wagenet** [comment]
@treycaliva AFAICT, this isn't something we can control. It's saying it can't find the right version of activesupport, but we're relying on Rubygems/Bundler to resolve gems.

**Comment by nicolasmlv** [comment]
I am using Rails 6.0.0.alpha and I have the same error

**Comment by wagenet** [comment]
@nicolasmlv thanks for sending your Gemfile. I tried running `bundle install` both with and without your `Gemfile.lock` and both worked fine. I too have Bundler 1.16.4 so I have no clue why it works for me and not for you. My best guess is some bug in Bundler, but I don't think its a Skylight problem directly.

**Comment by stormsilver** [comment]
We are having the same issue. We have forked rails temporarily to add some logging to help us diagnose an issue. When rails is installed from git, then this line causes issues:
https://github.com/skylightio/skylight-ruby/blob/2d15641f81a74f71d8f29ecf76e411a2a3873a5f/ext/extconf.rb#L17

The problem seems to be that the version of activesupport that comes along with rails from git is not active/valid/findable when `extconf.rb` is invoked and requires the skylight-core gem.

**Comment by wagenet** [comment]
@stormsilver is it possible for you to put together a reproduction that you can share? I hate to ask for that, but we still haven't been able to reproduce it on our end.

**Comment by stormsilver** [comment]
https://gist.github.com/stormsilver/01956047175df41ff1ca0e14c51bf633

The key is that you *must not* have activesupport (or skylight-*) installed already. I achieved this by creating a new gemset in rvm.

**Comment by wagenet** [comment]
@stormsilver thanks for the reproduction. I'm now seeing the issue locally and I will investigate further.

**Comment by wagenet** [comment]
It looks like the somewhat annoying workaround is to `gem install skylight-core` in advance. I'm continuing to look into whether there's a better option.

**Comment by wagenet** [comment]
Unfortunately, all indications are that this is a Bundler issue with git dependencies not being available when other non-git dependencies normally are. I'm not really sure what else we can do to resolve this.

**Comment by stormsilver** [comment]
It looks like you're requiring skylight-core so that you can use some of the utils from it. Can you remove the build-time dependency on skylight-core somehow?

**Comment by 97jaz** [comment]
Having the same problem (only on CI, not locally, which is odd but consistent with what a previous reported said). Installing the gem before doing a `bundle install` does not help.

**Comment by wagenet** [comment]
@stormsilver my apologies for missing your comment earlier. If this continues to be an issue for others this is something we can definitely consider.

@97jaz are you using an unreleased version of Rails as well?

**Comment by 97jaz** [comment]
@wagenet Yes, sorry for not mentioning that. There's no problem when using the rc1 release of rails 6, but when I switched over to using git in the Gemfile (to track the 6-0-stable branch), this started to be a problem on CircleCI (but not locally).

**Comment by wagenet** [comment]
@97jaz thanks for the information. It does indeed seem like this happens when `skylight-core` hasn't been installed before running `bundle install`. I'll bring this up with the rest of the team.

**Comment by zvkemp** [comment]
@97jaz After doing a deep dive into bundler and rubygems, we've determined that the issue is that there is a difference in how the `gem` method operates depending on whether or not bundler has been required. `-rbundler/setup` is notably absent from `RUBYOPT` when running extconf.rb, so we are using the plain rubygems-provided `gem`, which is unaware of gems installed via git by bundler. Therefore, it appears to rubygems that the `activesupport` dependency on `skylight-core` is unfulfilled. I have a fix in progress which should be released soon.

If you need a temporary solution immediately, you can also bundle skylight from github (e.g., `gem 'skylight', git: 'https://github.com/skylightio/skylight-ruby', ref: 'v4.1.0'`).

**Comment by zvkemp** [comment]
Fixed in [Skylight 4.1.1](https://rubygems.org/gems/skylight/versions/4.1.1)

**Comment by mendokusai** [comment]
Hi there, just wanted to ping in. I also had this trouble trying to bundle `skylight v 4.1.2` with ruby `2.4.4` and my issue wasn't skylight, but an out-of-date version of Xcode. As soon as I updated that, skylight bundled like a champ.
